src/frontend/src/scheduler/distributed/query.rs

```diff
@@ -242,9 +244,19 @@ impl QueryRunner {
                         self.query.query_id, stage_id
                     );
                     self.scheduled_stages_count += 1;
+                    leaf_stages.remove(&stage_id);
+                    if leaf_stages.is_empty() {
```

Comment:
Sorry for late reply and review. I'm afraid we may need to revert this PR since there may exists some case, for example:
```
           HashJoin
 /                                 \
Exchange               TableScan2
/
TableScan1
```

In this case, we have two fragments:

```
TableScan1
```

```
         HashJoin
            /           \
      Exchange TableScan2
```

In this case fragment2 may still need to do table scan and we can't unpin.

Reply:
I think a safer way is to create an `has_table_scan` attribute in fragmenter and attach it to each plan fragment, and unpin snapshot after all plan fragments with table scan has been scheduled.

Reply:
cc @lmatz

Reply:
I see, I thought this fragment 2 is considered as a leaf stage. Let me fix it.